File: lecture_6/Lecture6. Source code/topic_tutorial_py/subpose.py
import rclpy
import time
from rclpy.node import Node
from turtlesim.msg import Pose

class TurtlesimPoseSubscriber(Node):
    def __init__(self):
        super().__init__('turtlesim_pose_subscriber')
        self.subscription = self.create_subscription(
            Pose,
            'turtle1/pose',
            self.pose_callback,
            10)
        self.subscription  # prevent unused variable warning

    def pose_callback(self, msg):
        self.get_logger().info(f'Pose: x={msg.x}, y={msg.y}, theta={msg.theta}')

def main(args=None):
    rclpy.init(args=args)
    turtlesim_pose_subscriber = TurtlesimPoseSubscriber()
    
    start_time = time.time()
    while rclpy.ok() and (time.time() - start_time) < 3.0:
        rclpy.spin_once(turtlesim_pose_subscriber, timeout_sec=0.1)

    turtlesim_pose_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: lecture_6/src/topic_tutorial_cpp/src/subpose.cpp
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"
#include "memory"

using std::placeholders::_1;
class SubPoseNode : public rclcpp::Node
{
public:
    SubPoseNode() : Node("subscribe_pose_node")
    {
        subscription_ = this->create_subscription<turtlesim::msg::Pose>
            ("turtle1/pose", 10, std::bind(&SubPoseNode::pose_callback, this, _1));
    }
private:
    void pose_callback(turtlesim::msg::Pose::SharedPtr msg)
    {
        double x = msg->x;
        double y = msg->y;
        double theta = msg->theta;

        RCLCPP_INFO(this->get_logger(), "Turtlebot Pose -X: %f, T: %f, Theta: %f", x, y, theta);
    }
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
};

int main(int argv, char *argc[])
{
    rclcpp::init(argv, argc);
    auto node = std::make_shared<SubPoseNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
}